java/avro-converter/src/main/java/com/microsoft/azure/schemaregistry/kafka/connect/avro/AvroConverter.java

```diff
@@ -0,0 +1,132 @@
+    this.avroData = new AvroData(new AvroDataConfig(configs));
+
+    TokenCredential tokenCredential = new ClientSecretCredentialBuilder()
+      .tenantId((String) this.avroConverterConfig.getProps().get("tenant.id"))
+      .clientId((String) this.avroConverterConfig.getProps().get("client.id"))
+      .clientSecret((String) this.avroConverterConfig.getProps().get("client.secret"))
+      .build();
+
+    if (schemaRegistryClient == null) {
+      schemaRegistryClient = new SchemaRegistryClientBuilder()
+      .fullyQualifiedNamespace(this.avroConverterConfig.getSchemaRegistryUrl())
+      .credential(tokenCredential)
+      .buildAsyncClient();
+    }
+
+    serializer = new SchemaRegistryApacheAvroSerializerBuilder()
+    .schemaRegistryAsyncClient(schemaRegistryClient)
+    .schemaGroup(this.avroConverterConfig.getSchemaGroup())
+    .buildSerializer();
+
+    deserializer = new SchemaRegistryApacheAvroSerializerBuilder()
+    .schemaRegistryAsyncClient(schemaRegistryClient)
+    .schemaGroup(this.avroConverterConfig.getSchemaGroup())
+    .buildSerializer();
+  }
+
+  public byte[] fromConnectData(String topics, Schema schema, Object value) {
+    return fromConnectData(topics, null, schema, value);
+  }
+
+  public byte[] fromConnectData(String topics, Headers headers, Schema schema, Object value) {
+    Object avroObject = avroData.fromConnectData(schema, value);
+    Object avroValue = avroObject instanceof NonRecordContainer
+          ? ((NonRecordContainer) avroObject).getValue()
+          : avroObject;
+    
+    try {
+      MessageWithMetadata message = serializer.serializeMessageData(
+        avroValue,
+        TypeReference.createInstance(MessageWithMetadata.class)
+      );
+
+      byte[] contentTypeBytes = message.getContentType().getBytes();
+      headers.add("content-type", contentTypeBytes);
+      
+      return message.getBodyAsBinaryData().toBytes();
+    } catch (SchemaRegistryApacheAvroException e) {
+      throw new DataException("Failed to serialize Avro data: ", e);
+    } catch (Exception e) {
+      throw e;
+    }
+  }
+
+  public SchemaAndValue toConnectData(String topic, byte[] value) {
+    return toConnectData(topic, null, value);
+  }
+
+  public SchemaAndValue toConnectData(String topic, Headers headers, byte[] value) {
+    String contentTypeString = "";
+    String schemaId = "";
+
+    try {
+      MessageWithMetadata message = new MessageWithMetadata();
+      message.setBodyAsBinaryData(BinaryData.fromBytes(value));
+
+      Header contentTypeHeader = headers.lastHeader("content-type");
+      if (contentTypeHeader != null) {
+        contentTypeString = new String(contentTypeHeader.value());
+        message.setContentType(contentTypeString);
+      }
+      
+      Object deserializedMessage = deserializer.deserializeMessageData(
+        message,
+        TypeReference.createInstance(this.avroConverterConfig.getAvroSpecificType())
```

Comment:
if we always expect user to give us the type, then we probably want to catch the missing type back in configure()